First draft of reference docs on durable state

debasishg opened the pull request:
<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please reference them here - but don't use `fixes` notation.
-->
References https://github.com/akka/akka/issues/30277

It's best to run `sbt akka-docs/paradox` to look at the final format since I have made some changes in the tab structure as well. I now have a separate tab for _Akka Persistence (Durable State)_. The previous tab on _Akka Persistence_ has been renamed to _Akka Persistence (Event Sourcing)_. I did this to conceptually separate the 2 implementations.

debasishg committed:
First draft of reference docs on durable state

debasishg committed:
Removed unused code

debasishg committed:
scalafmt

octonato commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+The default separator when concatenating the `entityTypeHint` and `entityId` is `|`, but a custom separator
+is supported.
+
+@@@ note
+
+The `|` separator is also used in Lagom's `scaladsl.PersistentEntity` but no separator is used
+in Lagom's `javadsl.PersistentEntity`. For compatibility with Lagom's `javadsl.PersistentEntity`
+you should use `""` as the separator.
+
+@@@
```
We don't need this notice since there is no durable state actors in Lagom. 

This notice is useful only for users migrating from Lagom to Akka Persistence Typed (EventSourcedBehavior).

debasishg replied:
removed

octonato commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+
+The command handler is a function with 2 parameters, the current `State` and the incoming `Command`.
+
+A command handler returns an `Effect` directive that defines what event or events, if any, to persist. 
+Effects are created using @java[a factory that is returned via the `Effect()` method] @scala[the `Effect` factory].
+
+The two most commonly used effects are: 
+
+* `persist` will persist one single event or several events atomically, i.e. all events
+  are stored or none of them are stored if there is an error
+* `none` no events are to be persisted, for example a read-only command
```
I know, it draft...

We are still mention events in these sentences.

debasishg replied:
changed ..

octonato commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+
+More effects are explained in @ref:[Effects and Side Effects](#effects-and-side-effects).
+
+In addition to returning the primary `Effect` for the command `EventSourcedBehavior`s can also 
```
```suggestion
In addition to returning the primary `Effect` for the command `DurableStateBehavior`s can also 
```

debasishg replied:
changed ..

octonato commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+Effects are created using @java[a factory that is returned via the `Effect()` method] @scala[the `Effect` factory]
+and can be one of: 
+
+* `persist` will persist one single event or several events atomically, i.e. all events
```
Persist state, not events.

debasishg replied:
changed ..

octonato commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+The side effects are executed sequentially, it is not possible to execute side effects in parallel, unless they
+call out to something that is running concurrently (for example sending a message to another actor).
+
+It's possible to execute a side effect before persisting the event, but that can result in that the
+side effect is performed but the event is not stored if the persist fails.
```
Same: state, not event

debasishg replied:
changed ..

octonato reviewed (commented):
Looking good. 

I know is still in draft. Nonetheless, I pointed out a few places where we are still talking about events instead of state.

debasishg committed:
PR review feedbacks

debasishg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+---
+project.description: Durable State with Akka Persistence enables actors to persist your state for recovery on failure or when migrated within a cluster.
```
changed ..

debasishg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,220 @@
+---
+# Durable State
+
+You are viewing the documentation for the new actor APIs, to view the Akka Classic documentation, see @ref:[Classic Akka Persistence](../persistence.md).
```
done ..

debasishg committed:
More docs, changed indexing to have a separate tab for Durable State

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,208 @@
+---
+project.description: Durable State with Akka Persistence enables actors to persist its state for recovery on failure or when migrated within a cluster.
+---
+# Durable State
```
Just like `typed/persistence.md#tagging` there should be a section about tagging durable state changes.

debasishg replied:
Yes, it's still in the works ..

debasishg replied:
Now added ..

debasishg committed:
Some compilation fixes, also making Effect covariant in State

debasishg committed:
More documentation ..

debasishg commented:
@RayRoestenburg @octonato Added a lot more documentation. The page on Persistence (Durable State) is now complete I also moved _Persistence (Durable State)_ to a separate tab and renamed _Persistence_ tab to _Persistence (Event Sourcing)_. Will be good to have a review.

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+  scope2=test
+}
+
+You also have to select journal plugin and optionally snapshot store plugin, see 
```
refers to journal and snapshots, should only refer to durable state store

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+## Introduction
+
+This model of Akka Persistence enables a stateful actor / entity to store the full state afer processing each command instead of using event sourcing. This reduces the conceptual complexity and can be a handy tool for simple use cases. Very much like a CRUD based operation, the API is conceptually simple - a function from current state and incoming command to the next state which replaces the cur ...
```
```suggestion
This model of Akka Persistence enables a stateful actor / entity to store the full state after processing each command instead of using event sourcing. This reduces the conceptual complexity and can be a handy tool for simple use cases. Very much like a CRUD based operation, the API is conceptually simple - a function from current state and incoming command to the next state which replaces the current state in the database. 
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
````diff
@@ -0,0 +1,391 @@
+(State, Command) => State
+```
+
+The current state is always stored in the database. Akka Persistence would read that state and store it in memory. After processing of the command is finished, the new state will be stored in the database. The processing of the next command will not start until the state has been successfully stored in the database.
````
Maybe make it clear that the storage replaces with every change (no audit or history kept)

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+entities may have the same `entityId`. To create a unique `PersistenceId` the `entityId` should be prefixed
+with a stable name of the entity type, which typically is the same as the `EntityTypeKey.name` that
+is used in Cluster Sharding. There are @scala[`PersistenceId.apply`]@java[`PersistenceId.of`] factory methods
+to help with constructing such `PersistenceId` from an `entityTypeHint` and `entityId`.
```
I feel this part on entityId and persistenceId can use some work, or can refer to cluster sharding conceptual docs.

RayRoestenburg replied:
Like for instance you have to assume entityTypeHint has something to do with the entity type, or read up on it somewhere else. Do we have other docs that explain this conceptually so we can refer to it?

debasishg replied:
It does refer to the cluster sharding document just above this part

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+durabe state store.
+
+@ref:[Cluster Sharding](cluster-sharding.md) is typically used together with `DurableStateBehavior` to ensure
+that there is only one active entity for each `PersistenceId` (`entityId`).
```
The `entityId` is not introduced, so it's not clear what it means for a new reader in this sentence?

debasishg replied:
rephrased

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+@ref:[Cluster Sharding](cluster-sharding.md) is typically used together with `DurableStateBehavior` to ensure
+that there is only one active entity for each `PersistenceId` (`entityId`).
+
+The `entityId` in Cluster Sharding is the business domain identifier of the entity. The `entityId` might not
```
'business domain identifier' does not say much, and in this case I guess it is more relevant that this entityId is unique across the cluster?

debasishg replied:
rephrased

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+The default separator when concatenating the `entityTypeHint` and `entityId` is `|`, but a custom separator
+is supported.
+
+The @ref:[Persistence example in the Cluster Sharding documentation](cluster-sharding.md#persistence-example)
```
Maybe this can be pulled more to the beginning of this section. So we can just state, in cluster sharding, there are some more details about uniqueness, please see ...

debasishg replied:
didn't want to disturb the structure as it's the same in the other (event sourcing) document.

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+illustrates how to construct the `PersistenceId` from the `entityTypeKey` and `entityId` provided by the
+`EntityContext`.
+
+A custom identifier can be created with `PersistenceId.ofUniqueId`.  
```
Why do I need to know this here?

debasishg replied:
This is from the event sourcing document .. remove ?

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+* `persist` will persist state atomically, i.e. all state will be stored or none of them are stored if there is an error
+* `none` no state to be persisted, for example a read-only command
+
+More effects are explained in @ref:[Effects and Side Effects](#effects-and-side-effects).
```
Maybe nice before this just to see an example of how this is used.

debasishg replied:
again, for uniformity with the other document I kept it this way ..

debasishg replied:
Thought of adding an example here. But found that the running example is just a few lines away - hence no point repeating it.

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+More effects are explained in @ref:[Effects and Side Effects](#effects-and-side-effects).
+
+In addition to returning the primary `Effect` for the command `DurableStateBehavior`s can also 
```
```suggestion
In addition to returning the primary `Effect` for the command, `DurableStateBehavior`s can also 
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+Effects are created using @java[a factory that is returned via the `Effect()` method] @scala[the `Effect` factory]
+and can be one of: 
+
+* `persist` will persist state atomically, i.e. all states are stored or none of them are stored if there is an error
```
Not sure about all states, is it not the case that you always persist one state (with durable state)?

debasishg replied:
corrected ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+* `thenUnstashAll` process the commands that were stashed with @scala[`Effect.stash`]@java[`Effect().stash`]
+* `thenReply` send a reply message to the given `ActorRef`
+
+In the example below, we use a different constructor of `DurableStateBehavior.withEnforcedReplies`, which creates
```
What does it mean that the reply is not forgotten? why would a reply be forgotten without this? (trying to read this as a new reader, it's not very clear to me, it assumes some knowledge of what can go wrong if I don't use it.)

RayRoestenburg replied:
Oh, maybe this means that you could forget to define a reply, as the person writing this, and by using this constructor, you can't define a command that does not have a reply. 

You could misread this as if Akka forgets replies. So maybe make it clear that this is related to the case where you as a developer want to ensure this.

debasishg replied:
refactored this part quite a bit ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+Java
+:  @@snip [DurableStatePersistentBehaviorTest.java](/akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/DurableStatePersistentBehaviorTest.java) { #effects }
+
+Most of the time this will be done with the `thenRun` method on the `Effect` above. You can factor out
```
Not very clear what this refers to, better to repeat something like 'It is useful to define side effects with `thenRun` ...'

debasishg replied:
rephrased

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+The side effects are executed sequentially, it is not possible to execute side effects in parallel, unless they
+call out to something that is running concurrently (for example sending a message to another actor).
+
+It's possible to execute a side effect before persisting the state, but that can result in that the
```
```suggestion
It's possible to execute a side effect before persisting the state, but that can cause that the
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+call out to something that is running concurrently (for example sending a message to another actor).
+
+It's possible to execute a side effect before persisting the state, but that can result in that the
+side effect is performed but the state is not stored if the persist fails.
```
```suggestion
side effect is performed but that the state is not stored, if the persist fails.
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+In a use case where the number of persistent actors needed is higher than what would fit in the memory of one node or
+where resilience is important so that if a node crashes the persistent actors are quickly started on a new node and can
+resume operations @ref:[Cluster Sharding](cluster-sharding.md) is an excellent fit to spread persistent actors over a
+cluster and address them by id.
```
Would be great if this sentence can be redone in a few sentences.

RayRoestenburg replied:
(I added a suggestion)

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+## Cluster Sharding and DurableStateBehavior
+
+In a use case where the number of persistent actors needed is higher than what would fit in the memory of one node or
```
```suggestion

@ref:[Cluster Sharding](cluster-sharding.md) is an excellent fit to spread persistent actors over a
cluster. It makes it possible to have more persistent actors exist in the cluster than what would fit in the memory of one node. 
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+## Cluster Sharding and DurableStateBehavior
+
+In a use case where the number of persistent actors needed is higher than what would fit in the memory of one node or
+where resilience is important so that if a node crashes the persistent actors are quickly started on a new node and can
```
Cluster sharding improves the resilience of the cluster, if a node crashes the persistent actors are quickly started on a new node and can resume operations.

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+resume operations @ref:[Cluster Sharding](cluster-sharding.md) is an excellent fit to spread persistent actors over a
+cluster and address them by id.
+
+The `DurableStateBehavior` can then be run as with any plain actor as described in @ref:[actors documentation](actors.md),
```
```suggestion
The `DurableStateBehavior` can then be run as any plain actor as described in @ref:[actors documentation](actors.md),
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+cluster and address them by id.
+
+The `DurableStateBehavior` can then be run as with any plain actor as described in @ref:[actors documentation](actors.md),
+but since Akka Persistence is based on the single-writer principle the persistent actors are typically used together
```
```suggestion
but since Akka Persistence is based on the single-writer principle, the persistent actors are typically used together
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+The `DurableStateBehavior` can then be run as with any plain actor as described in @ref:[actors documentation](actors.md),
+but since Akka Persistence is based on the single-writer principle the persistent actors are typically used together
+with Cluster Sharding. For a particular `persistenceId` only one persistent actor instance should be active at one time.
```
mixing entity with actor, though entity is not explained. Maybe just say "ensures that there is only one active actor instance for each id' ?

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+## Accessing the ActorContext
+
+If the @apidoc[DurableStateBehavior] needs to use the @apidoc[typed.*.ActorContext], for example to spawn child actors, it can be obtained by
+wrapping construction with `Behaviors.setup`:
```
`PersistenceId.ofUniqueId("myPersistenceId")` is not clear, what this does. I guess it expresses that you know, as a user that this is unique, it does not automatically make it unique?

debasishg replied:
yeah .. it assumes that the id will be unique ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+## Changing Behavior
+
+After processing a message, actors are able to return the `Behavior` that is used
```
Maybe simpler to just state something like : Unlike regular actors, persistent actors cannot return a different behavior for a next message.

RayRoestenburg replied:
I would prefer it if this is written, focused on how it should be done, and then a note saying "while reading this you might have wondered why we did not use ..., this is not allowed because' or something and explaining there why the behavior changing cannot work with akka persistence.

debasishg replied:
refactored ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+The reason a new behavior can't be returned is that behavior is part of the actor's
+state and must also carefully be reconstructed during recovery from the persisted state. This would imply
+that the state needs to be encoded such that the behavior can also be restored from it. 
+That would be very prone to mistakes and thus not allowed in Akka Persistence.
```
```suggestion
That would be very prone to mistakes , which is why it is not allowed in Akka Persistence.
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+given to the `forStateType` of the `CommandHandlerBuilder` and the match cases in the builders.]
+@scala[The command handler to process each command is decided by first looking at the state and then the command.
+It typically becomes two levels of pattern matching, first on the state and then on the command.]
+Delegating to methods is a good practice because the one-line cases give a nice overview of the message dispatch.
```
Add here some reference to `addPost` and other methods to be clear

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+Java
+:  @@snip [BlogPostEntityDurableState.java](/akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/BlogPostEntityDurableState.java) { #behavior }
+
+This can be taken one or two steps further by defining the command handlers in the state class as
```
```suggestion
This can be refactored one or two steps further by defining the command handlers in the state class as
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+
+Since this is such a common pattern there is a reply effect for this purpose. It has the nice property that
+it can be used to enforce that replies are not forgotten when implementing the `DurableStateBehavior`.
```
```suggestion
it can be used to enforce that you do not forget to specify replies when implementing the `DurableStateBehavior`.
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+These effects will send the reply message even when @scala[`DurableStateBehavior.withEnforcedReplies`]@java[`DurableStateBehaviorWithEnforcedReplies`]
+is not used, but then there will be no compilation errors if the reply decision is left out.
+
+Note that the `noReply` is a way of making conscious decision that a reply shouldn't be sent for a specific
```
```suggestion
Note that the `noReply` is a way of making a conscious decision that a reply shouldn't be sent for a specific
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+is not used, but then there will be no compilation errors if the reply decision is left out.
+
+Note that the `noReply` is a way of making conscious decision that a reply shouldn't be sent for a specific
+command or the reply will be sent later, perhaps after some asynchronous interaction with other actors or services.
```
```suggestion
command or that a reply will be sent later, perhaps after some asynchronous interaction with other actors or services.
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+## Tagging
+
+Persistence allows you to use tags in persistence query.
```
This needs a little more explanation IMHO

RayRoestenburg replied:
The java example seems to point to something else.

debasishg replied:
added some explanation explaining tags. The Java example points to the method which overrides `def tag: String` in the javadsl .. Now added the class definition as well, so that we know which class are we talking about

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-durable-state.md:
```diff
@@ -0,0 +1,391 @@
+
+## Wrapping DurableStateBehavior
+
+When creating an `DurableStateBehavior`, it is possible to wrap `DurableStateBehavior` in
```
```suggestion
When creating a `DurableStateBehavior`, it is possible to wrap `DurableStateBehavior` in
```

debasishg replied:
done ..

RayRoestenburg commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+@java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter
+will then be `null` for the first commands until the first non-null state has been persisted 
+It's possible to use `Optional` instead of `null` but that results in rather much boilerplate
+to unwrap the `Optional` state parameter and therefore `null` is probably preferred. The following example
```
```suggestion
to unwrap the `Optional` state parameter and therefore `null` is preferred. The following example
```

debasishg replied:
done ..

RayRoestenburg reviewed (commented):
Did a first pass on the docs, already looks quite good, I made some suggestions.

debasishg committed:
PR review feedbacks

debasishg commented:
@ray Incorporated almost all suggestions .. kept a couple unaltered to maintain coherence with the event sourcing doc

rstento commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+
+Sometimes it's not desirable to use a separate state class for the empty initial state, but rather treat that as
+there is no state yet.
+@java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter
```
```suggestion
@java[You can use `null` can as the `emptyState`, but be aware of that the `state` parameter
```

thothmoses replied:
I think the second 'can' is unneeded - (after 'null')
@java[You can use `null` as the `emptyState`, but be aware of that the `state` parameter

debasishg replied:
done ..

rstento commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+Sometimes it's not desirable to use a separate state class for the empty initial state, but rather treat that as
+there is no state yet.
+@java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter
+will then be `null` for the first commands until the first non-null state has been persisted 
```
```suggestion
will be `null` until the first non-null state has been persisted 
```

debasishg replied:
done ..

rstento commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+there is no state yet.
+@java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter
+will then be `null` for the first commands until the first non-null state has been persisted 
+It's possible to use `Optional` instead of `null` but that results in rather much boilerplate
```
```suggestion
It's possible to use `Optional` instead of `null`, but that requires extra boilerplate
```

debasishg replied:
done ..

rstento commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+@java[`null` can then be used as the `emptyState`, but be aware of that the `state` parameter
+will then be `null` for the first commands until the first non-null state has been persisted 
+It's possible to use `Optional` instead of `null` but that results in rather much boilerplate
+to unwrap the `Optional` state parameter and therefore `null` is preferred. The following example
```
```suggestion
to unwrap the `Optional` state parameter. Therefore, use of  `null` is simpler. The following example
```

debasishg replied:
done ..

rstento reviewed (approved):
Looks good, @debasishg! I left a few suggestions.

thothmoses commented on akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/AccountExampleWithNullDurableState.java:
```diff
@@ -0,0 +1,218 @@
+import java.math.BigDecimal;
+
+/**
+ * Bank account example illustrating: - different state classes representing the lifecycle of the
+ * account - null as emptyState - event handlers that delegate to methods in the state classes -
+ * command handlers that delegate to methods in the DurableStateBehavior class - replies of various
+ * types, using DurableStateBehaviorWithEnforcedReplies
```
```suggestion
 * This bank account example illustrates the following:  different state classes representing the lifecycle of the
 * account, null as emptyState, event handlers that delegate to methods in the state classes,
 * command handlers that delegate to methods in the DurableStateBehavior class, and replies of various
 * types that use DurableStateBehaviorWithEnforcedReplies
```

debasishg replied:
done ..

thothmoses commented on akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithCommandHandlersInDurableState.scala:
```diff
@@ -0,0 +1,135 @@
+import akka.serialization.jackson.CborSerializable
+
+/**
+ * Bank account example illustrating:
```
```suggestion
 * This bank account example illustrates:
```

debasishg replied:
done ..

thothmoses commented on akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithOptionDurableState.scala:
```diff
@@ -0,0 +1,126 @@
+import akka.serialization.jackson.CborSerializable
+
+/**
+ * Bank account example illustrating:
```
```suggestion
 * This bank account example illustrates:
```

debasishg replied:
done ..

thothmoses commented on akka-docs/src/main/paradox/typed/index-persistence-durable-state.md:
```diff
@@ -0,0 +1,14 @@
+---
+project.description: Event Sourcing with Akka Persistence enables actors to persist your events for recovery on failure or when migrated within a cluster.
```
```suggestion
project.description: Event Sourcing with Akka Persistence enables actors to persist your events. This persistence is for recovery on failure, or when migrated within a cluster.
```

thothmoses commented on akka-docs/src/main/paradox/typed/index-persistence-durable-state.md:
```diff
@@ -0,0 +1,14 @@
+---
+project.description: Event Sourcing with Akka Persistence enables actors to persist your events for recovery on failure or when migrated within a cluster.
```
```suggestion
project.description: Use of Event Sourcing with Akka Persistence enables actors to persist your events. This persistence is used for recovery on failure, or when migrating within a cluster.
```

debasishg replied:
actually it was a mistake to describe this as event sourcing (copied it from the other doc and forgot to change). It's not. Changed the text to the following:

> Durable state with Akka Persistence enables actors to persist the latest version of the state. This persistence is used for recovery on failure, or when migrating within a cluster.

thothmoses commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+@@@ div { .group-scala }
+## Command handlers in the state
+
+We can take the previous bank account example one step further by handling the commands in the state too.
```
```suggestion
We can take the previous bank account example one step further by handling the commands within the state as well.
```

debasishg replied:
done ..

thothmoses commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+Scala
+:  @@snip [AccountExampleWithCommandHandlersInDurableState.scala](/akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithCommandHandlersInDurableState.scala) { #account-entity }
+
+Notice how the command handler is delegating to `applyCommand` in the `Account` (state), which is implemented
```
```suggestion
Take note of how the command handler is delegating to `applyCommand` in the `Account` (state), which is implemented
```

debasishg replied:
done ..

thothmoses commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+
+## Optional initial state
+
+Sometimes it's not desirable to use a separate state class for the empty initial state, but rather treat that as
+there is no state yet.
```
```suggestion
Sometimes it's not desirable to use a separate state class for the empty initial state, but rather act as if
there is no state yet.
```

debasishg replied:
done ..

thothmoses commented on akka-docs/src/main/paradox/typed/persistence-style-durable-state.md:
```diff
@@ -0,0 +1,32 @@
+It's possible to use `Optional` instead of `null` but that results in rather much boilerplate
+to unwrap the `Optional` state parameter and therefore `null` is preferred. The following example
+illustrates using `null` as the `emptyState`.]
+@scala[`Option[State]` can be used as the state type and `None` as the `emptyState`. Pattern matching
+is then used in command handlers at the outer layer before delegating to the state or other methods.]
```
```suggestion
@scala[`Option[State]` can be used as the state type and `None` as the `emptyState`. Then pattern matching
is used in command handlers at the outer layer, before delegating to the state or other methods.]
```

debasishg replied:
done ..

thothmoses reviewed (changes requested):
a few tiny changes

debasishg commented on akka-docs/src/main/paradox/typed/index-persistence.md:
```diff
@@ -2,7 +2,7 @@
 project.description: Event Sourcing with Akka Persistence enables actors to persist your events for recovery on failure or when migrated within a cluster.
```
Actually Akka Persistence is a broader topic, in which, Event Sourcing is one of the ways to achieve persistence. So, they cannot really use Event Sourcing without using Akka Persistence. But I agree something like "Use of Akka Persistence with Event Sourcing enables .." sounds better. 

Changed ..

debasishg committed:
PR review feedbacks

debasishg commented:
@rstento @thothmoses Addressed all feedbacks ..

debasishg commented:
@RayRoestenburg @thothmoses @rstento This PR has already grown too big. I still plan to add some more stuff on Query, CQRS and (maybe) Testing. Can we merge this PR and add the rest in a separate PR ?

Brian - All changes that u suggested have been done. Once I get a green tick from you I can merge this PR ..

RayRoestenburg reviewed (approved):
LGTM
